### ISSUE TITLE
Slow speeds?

### ISSUE THREAD

**Issue by storedcc** [steamcache/steamcache#15]
Hey there!

I've encountered some very slow speeds with the steamcache. Tried it out some months ago, but it doesn't work as well as it should. 

The main problem is that it barely goes over 80 to 100 mbit/s. It has to be said that even full cached games aren't delivered with at least 200 - 300mbit on a Gigabit lan connection. The Disk i/o as well as the CPU utilization is very low so I am a bit confused, but maybe someone has an idea. Thanks in advance :)

**Comment by Azothyran** [comment]
From personal experience, I get from 7-10MB/s average download speeds through Steamcache while downloading a game that is NOT already in the cache and and an average of 13.7MB/s without Steamcache intercepting the data at all. The connection I'm paying for from my ISP caps out at 14.5MB/s.

After a game is cached, I achieve somewhere between 30-45MB/s through my network from the Steamcahce box if I reinstall it on the same or different machines (it's very consistent even on differing hardware).

![steamcache working cropped](https://cloud.githubusercontent.com/assets/26119600/24589804/9b462426-17a6-11e7-84b2-3525aacfccb7.jpg)

I have my Steamcache box running on Ubuntu Server as a Hyper-V guest with Windows as a host OS, and only gave the VM 1GB of RAM since it didn't seem to need much more than that. I passed a 7200RPM 2TB drive directly to the VM as a second drive mounted to a folder in the Ubuntu Server VM in which Steamcache is dumping its depot data directly to the disk, and Windows, the Steamcache Box, a PiHole DNS server (running as a second Ubuntu Server VM) running together on a single Gigabit link. I have pfSense in a third VM working as my router / firewall through a dedicated addon dual NIC.

Long story short, my setup is a bit crazy and probably not ideal, but despite that, I slightly exceed the rates you're reporting in my personal case.

**Comment by storedcc** [comment]
Im trying to be nice here - That's fine for you, but doesn't explain nor solve my issue ^^

I'm willing to give out any information needed to get this fixed, I just can't explain it, since the system utilization is so low..

**Comment by Azothyran** [comment]
@storedcc  I was merely giving a point of reference to corroborate your findings so the devs can chime in with whether these speeds are the norm.

**Comment by storedcc** [comment]
Hm. Okay... I may also be able to provide some more info at this point:

AMD Phenom(tm) II X4 840 Processor
8GB Ram
SAMSUNG SpinPoint F3 / SAMSUNG HD103SJ/ 7200 rpm

CPU utilization: [https://screen.stored.cc/2_7_02-04-2017_20-51-32.png](url)
i/o utilization (iotop): [https://screen.stored.cc/2_7_02-04-2017_20-52-27.png](url)
Network utilization: [https://screen.stored.cc/2_4_02-04-2017_20-52-49.png](url)

**Comment by storedcc** [comment]
It somehow fixed itself, I am now getting the speeds I was used to

**Comment by astrolox** [comment]
I see you've closed this already, but I wanted to add some comments which may be of help to anyone else with a similar question ...


# Quick answers ...

Is it just slow? No.

It it "tuned" for performance? Yes, mostly.

Does running in docker make it slower? No (I guess technically it has to because more work involved, but it's so insignificant you can't measure it)

Is it perfect? Almost definitely not. Let us know if you see something that should be changed to get better performance.


## Troubleshooting

As to finding the problem, checking disk IO, CPU utilisation, and RAM utilisation are the obvious things. Any of these being maxed out probably indicates you've found the bottleneck.

Note that Linux uses up unused RAM to cache files stored on disk (see http://www.linuxatemyram.com/). So you want to have lots and lots of unused RAM to get good performance. Check your RAM utilisation. ... Maybe there are other processes on the same server which are using RAM and stopping it from being used for disk caching.

Check the specifications of your equipment. For example the above mentioned hard drive maxes out at around the ~ 140 MB/s according to Toms Hardware benchmarks. That's assuming absolute best conditions, real performance will be lower. So, I'd expect this disk to be unable to saturate a standard 1 Gbps network card, unless you get really lucky on which files steam needs, in what order, and when. 

Consider your use cases and test scenario ... Say you have several people downloading a 60 GB game but nothing else. Having more than 60 GB of RAM means that your cache will never have to wait for the disk. With less RAM having everyone start the download at the same time will give a similar level of performance. However having every download different games will mean lots of disk use.

Also please be careful with comparing numbers ... 100 Megabits per second (Mbps) is 12.5 Megabytes per second (MB/s).


### Don't assume a local cache will always be faster than the Internet servers

Obviously if you have fast internet and are setting up a steam cache on a local server then it is very possible that you will make accessing steam content slower. The steam servers on the Internet are all pretty good servers (in big clusters of multiple servers). Basically they're fast so it's really only your Internet that limits your download speed. With your own server in the mix you're adding more bottlenecks.

However if you have lots of people wanting to download the same game and your Internet connection can't cope - then a steam cache will help you. This is the scenario we wanted to solve.


## Our experience

I can confirm that we (the original authors of this image) use it at LAN parties with in excess of several hundred people and see throughput much higher than the numbers quoted here. We use dedicated hardware, not virtual machines, specifically to get performance. For the reasons mentioned above we don't run other workloads on the same server.

A few years ago we used a server with 16GB of RAM and mirrored SAS drives for 300 clients. Observed peek performance was around 280 Mbps (35 MB/s). Our conclusion was to add more RAM.

I don't have up to date statistics available to me at the moment, I'll ask about getting some posted.

**Comment by mintopia** [comment]
I'd like to build upon the comments from @astrolox as I have some of the stats from our previous event and also go into some of the history and explain some of the decisions/things we've learnt since doing this.

### SteamCache History

I'm the technical manager for a 480 person LAN party in the UK. Our bandwidth for the event is very limited so the main aims for me of using various caches are:

- Reduce overall bandwidth usage
- Aim to make game installs quick
- Deal with download/update spikes caused by a publisher updating a game
- Reliability - it needs to work all the time, with no downsides.

Collectively we've been working on LAN caches for a few years. Originally Multiplay started it but their setup was awkward to reproduce and whilst flexible, was very complex to keep running. @astrolox and I worked on an Ubuntu version that was simpler and to the point.

As Docker has become the best way to run and deploy things like this, we turned it into a Docker image and started using that instead. At this point we decided to publish it publically so other people can contribute and use it.

### Proxy Cache vs Proxy Store

One change we made from the original MPUK version was to use Proxy Store. This would create a mirror of the content server on the cache and allow us to easily manage and maintain the cache. This came out of a neccessity of not having the huge amounts of storage Multiplay built theirs with.

The advantage is that it allows us to see everything cached, sorted by depot. The disadvantage is that if a chunk gets corrupt when downloaded/stored then the cache won't obey the client requesting it to be re-downloaded. We saw occasions at events where a client would be stuck downloading a chunk from the cache and the only fix was to delete the cached file and try again. This break with reliability caused me to switch to using Proxy Cache.

It's harder to maintain but fixes that reliability problem immediately. Since then I've had no issues.

### Multiple Cache Instances

Caches store things based on the cache key. This is normally done using the hostname of the request and the path and filename. However the services we're caching have Content Delivery Networks and serve up the content from multiple hostnames. This means you have to exclude the hostname from the cache key, otherwise you'd end up with valve205.cs.steampowered.com/myfile.txt and valve204.cs.steampowered.com/myfile.txt being stored and cached separately.

Because the hostname is not (and should not) be in the cache key; this means you should not have content from multiple different sources going to the same cache server. It's unlikely there'll be a collision, but it could happen. myfile.txt could be downloaded for a Steam Content Cache and then served up when requested by Origin.

The solution is to just run multiple cache instances, something Docker makes easy, add multiple IPs to the cache host and bind each instance to a separate IP. As of the small event last weekend, I have caches for the following:

 - Steam
 - Origin
 - Blizzard/Battle.net
 - Riot/League of Legends
 - Frontier
 - UPlay
 - Windows Update
 - Steam Workshop

### Docker vs Non-Docker vs Different Networking Modes

There is an overhead to using Docker, however it is very very minimal. I have tested under heavy 400 person load running the cache natively and running it under Docker. There was no noticeable difference. I run it in normal bridged networking mode and it's fine and that is never the bottleneck.

### Hardware and Specification

The key thing for the cache is RAM. We have 32GB of RAM in our cache server. I want more. As @astrolox explained, Linux will cache files in RAM in order to prevent reads from disk. Our Steam Cache docker container sits at around 30GB RAM consumed in normal use. If you're running it in a VM give all the spare RAM to your cache server. The RAM is what makes it quick.

The fixed disk underpinning this is 2 x 4TB Western Digital Reds in Striped RAID. I would have more disks and faster disks but the entire event core rack is 8U and I don't have space for a server that could fit more disks.

As mentioned, it's run entirely on dedicated hardware. No virtualisation, just bare metal Ubuntu 16.04, Docker and Rancher for orchestration.

We used to have 16GB of RAM and found we were bottlenecking hard on disk I/O as it was doing lots of random reading across the entire disk. CPU usage for the cache was 100% and load average was up at 16+. Since upgrading the RAM, we no longer have an issue with disk I/O.

### Network Performance

We used to have a 10gbps NIC in our old cache, it's currently on 1gbps, but I'm going to upgrade it to 4gbps (4 x 1gbps LACP link) when I acquire a new Intel 4 port NIC.

Last event with 480 people, we saw a peak bandwidth on the cache graphs of 430 mbps. However, this is based on 5 minute averages and it spiked up to 1gbps on a few occasions. We have 4 gbps to the aggregation switches, 1 gbps to the desk switches and 100 mbps to the desk for each customer. If we offered gigabit to the customers, we'd see higher values. But there's notes below on the usage patterns we saw for the cache.

The cache had access to 30mbps of download bandwidth and it used it literally continuously for 72 hours. We downloaded a total of 1.24 TiB from the Internet and uploaded 3.09 TiB to the LAN. The cache was starting from a mostly empty state as only the most popular games had been primed.

### Usage Patterns

There's a few usage patterns for the cache that we've seen:

1. A patch is released for a game like Dota2. All the clients on the LAN attempt to download it. Generally these are small updates of maybe a few hundred MB. They get cached quickly and served up mostly from RAM as everyone is getting the update at the same time. Because of the size of the update, people download it quickly and there's no network bottleneck.
2. A customer decides to play some random game. If it's on the cache, they get it at whatever speed the cache can read it from disk and send it. Speed will mostly be dependent on disk I/O. If lots of people are doing this, then you would need to improve disk performance to improve it. If it's not on the cache, it'll be downloaded at whatever speed we can allocate to the cache.
3. A group of customers decide to all play the same random game. This doesn't take much longer than scenario 2, since the bottleneck is disk I/O, it only needs to be read once from disk.

By upgrading the RAM in our main cache server, we managed to solve the disk I/O issues we were seeing originally.

The key thing with the performance is that there's a lot of things that can cause it to vary. At a large scale however, the cache works incredible well for achieving the original goals.

### Improvements

There are some definite improvements I'd like to make and @kixelated has done some of these and I'd love to pull them back into the generic image (I don't use the steamcache image).

- Fix range requests and possible problems with some games. I've personally not seen it in our setup, but totally accept that it could be an issue.
- Better documentation on how to set it up.
- Improved metrics and reporting. There's been some great work done on this by other people. I'd love to bring that into a companion image that can potentially analyse the logs.
- Improved logging. Sending the logs to docker properly is the correct way to do it, as long as it can work nicely with a toolchain for monitoring the data and cache properly, it'd be great.

### Key Points

The key points to takeaway for me would be:

- The cache is best for when you have multiple users wanting to download games. If there's only you testing, you won't see the full benefit as there are compromises made to support the use case of multiple users.
- RAM. More RAM. Take the amount of RAM you think you want, and double it. You want as much data coming from RAM as possible. Going back to disk, even SSDs, is slow.
- Keep different cache namespaces separate. Do not use the same cache for Blizzard, Steam, Origin, etc. You risk cache collisions and honestly the overhead for multiple nginx processes is worth the separation. Docker makes it *very* easy to do this. just add a new IP to the network interface/a new network interface, add DNS records and run a new instance of the cache pointing to separate storage.

**Comment by astrolox** [comment]
wow, thanks Jess. Lots of good info there.

I think maybe need to move find a better home to put some of the comments not relating to performance, so as not to confuse anyone who finds this issue in the future.

**Comment by kixelated** [comment]
Hey @mintopia, thanks for the writeup. I finally got around to rebasing my changes on steamcache/generic and making a pull request. I really hate calling my repo `generic` but I'll survive...

https://github.com/steamcache/generic/pull/15

The relevant improvements I've added:
- Differentiate requests by the Host header.
- Treat services differently to fix bugs with League/Battle.net/Steam.
- Include the $server_name in the proxy_cache_key (avoids collisions). You won't need to have separate instances any longer.
- Log to stdout instead of /data/log


I've got some more tasks in the works, like adding docker-compose support and removing some hard-coded variables. Let me know if you have any more suggestions!

**Comment by mintopia** [comment]
Hi @kixelated,

Thanks for the PR, I've replied in some comments to the PR. My issue is purely around the original purpose of the generic image and making it un-generic. I agree those fixes need to be in and we need the differentiation of upstream services, so it's working out the best approach to it.

Agree also with the logging to stdout, although would like the option to also log to file as well for people who are already integrated with file-based logging.

We'll pick it up over in steamcache/generic#15 though.

**Comment by gelsas** [comment]
Hello,

before using steamcache I had download speeds from 45-50MB/s .

Now with steamcache (run in docker) the initial download speed is at max 8-9 MB/s per second. 

What could the issue be here?

**Comment by RavenX8** [comment]
I'm having the same issue. I've tried a speed test from the cache host and it can hit 600Mbps, but when download games using the cache server I only hit 5 - 20 Mbps. Without the server I get 200-400Mbps. I know it isn't disk speed since I can download and write a 1000MB file at 264Mbps.

```raven@lancache:/lancache$ sudo wget http://speedtest-ny.turnkeyinternet.net/1000mb.bin
--2018-09-19 19:03:25--  http://speedtest-ny.turnkeyinternet.net/1000mb.bin
Resolving speedtest-ny.turnkeyinternet.net (speedtest-ny.turnkeyinternet.net)... 173.198.210.94, 2604:7c00::2
Connecting to speedtest-ny.turnkeyinternet.net (speedtest-ny.turnkeyinternet.net)|173.198.210.94|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1073741824 (1.0G) [application/octet-stream]
Saving to: ‘1000mb.bin.1’

1000mb.bin.1                                                100%[========================================================================================================================================>]   1.00G  46.4MB/s    in 24s

2018-09-19 19:04:03 (42.9 MB/s) - ‘1000mb.bin.1’ saved [1073741824/1073741824]
```

CPU Usage is at 1% while downloading and I have 5.5 GB of free unused ram on the host system.

Any ideas on why this would be?

**Comment by unk1nd** [comment]
Seems that the blizzard cache is missing some nodes, seems to bypass a lot of the data.
Also, as in the write up, I am running this on 128GB ram server with RAID 10 setup with 25x 600GB SAS disks. So Steam is caching like a god, to bad it bypass on the blizzard cache and gives the download in 8 MB\s instead of giving the traffic 60MB\s as it should on my connection.

**Comment by dezmaeth** [comment]
> Hello,
> 
> before using steamcache I had download speeds from 45-50MB/s .
> 
> Now with steamcache (run in docker) the initial download speed is at max 8-9 MB/s per second.
> 
> What could the issue be here?

i have the same issue, tried everything

**Comment by nexusofdoom** [comment]
What was the fix? I am having issues as of Dec 2018.

**Comment by pkloodt** [comment]
Same problem here! Steam cache running as Docker Containers on Ubuntu Desktop 18.04. 
Hardwarewise it's an Intel Pentium G4560 with 20GB RAM and 6x4TB 5400RPM Drives in ZFS Raid-10, but also tested with a single 1TB 7200RPM Drive with ext4.

Before steamcache I had download rates at 35-40MB/s, with steamcache they're max at 3-5MB/s. 

htop and iotop show that the system is not under high load, it just doesn't seem to leverage all it's potential...
